File: JS/Utilities/newProblem.js
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');


const mainTemplate = fs.readFileSync(path.join(__dirname, '../templates', 'main.js'), 'utf-8');

const testTemplate = fs.readFileSync(path.join(__dirname, '../templates', 'main.test.js'), 'utf-8');

function createFileFromTemplate(filePath, template, problem) {
    const content = template.replace(/{{ProblemName}}/g, problem.problemName);
    fs.writeFileSync(filePath, content, 'utf8');
}

function main() {
    const args = process.argv.slice(2);
    if (args.length !== 2) {
        console.log('Usage: <program> <problem_name> <problem_number>');
        process.exit(1);
    }

    const problem = {
        problemName: args[0],
        problemNumber: args[1]
    };

    const dirPath = path.join(__dirname, '..', 'Problems', `${problem.problemNumber}_${problem.problemName}`);
    if (fs.existsSync(dirPath)) {
        console.log('Problem already exists');
        process.exit(1);
    }

    fs.mkdirSync(dirPath, { recursive: true });

    createFileFromTemplate(path.join(dirPath, 'main.js'), mainTemplate, problem);
    createFileFromTemplate(path.join(dirPath, 'main.test.js'), testTemplate, problem);

    console.log(`Problem ${problem.problemNumber}_${problem.problemName} created successfully in ${problem.problemNumber}_${problem.problemName}`);
}

main();

File: Go/Utilities/newProblem.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"text/template"
)

const goMainTemplate = `package main

import "fmt"

func main() {
    fmt.Println("Hello, {{.ProblemName}}!")
}
`

const goTestTemplate = `package main

import (
    "bytes"
    "io"
    "os"
    "testing"
)

func TestMain(t *testing.T) {
    // Save the original stdout
    originalStdout := os.Stdout
    r, w, _ := os.Pipe()
    os.Stdout = w

    // Call the main function
    main()

    // Capture the output
    w.Close()
    var buf bytes.Buffer
    io.Copy(&buf, r)
    os.Stdout = originalStdout

    // Check the output
    expected := "Hello, {{.ProblemName}}!\\n"
    if buf.String() != expected {
        t.Errorf("expected %q but got %q", expected, buf.String())
    }
}
`

type Problem struct {
	ProblemName   string
	ProblemNumber string
}

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Usage: <program> <problem_name> <problem_number>")
		os.Exit(1)
	}

	problem := Problem{
		ProblemName:   os.Args[1],
		ProblemNumber: os.Args[2],
	}

	dirPath := filepath.Join("../", fmt.Sprintf("%s_%s", problem.ProblemNumber, problem.ProblemName))
	if _, err := os.Stat(dirPath); !os.IsNotExist(err) {
		fmt.Println("Problem already exists")
		os.Exit(1)
	}

	if err := os.MkdirAll(dirPath, 0755); err != nil {
		fmt.Printf("Failed to create directory: %v\n", err)
		os.Exit(1)
	}

	createFileFromTemplate(filepath.Join(dirPath, "main.go"), goMainTemplate, problem)
	createFileFromTemplate(filepath.Join(dirPath, "main_test.go"), goTestTemplate, problem)

	fmt.Printf("Problem %s_%s created successfully in %s_%s\n", problem.ProblemNumber, problem.ProblemName, problem.ProblemNumber, problem.ProblemName)
}

func createFileFromTemplate(filePath, tmpl string, problem Problem) {
	t := template.Must(template.New("file").Parse(tmpl))
	file, err := os.Create(filePath)
	if err != nil {
		fmt.Printf("Failed to create file %s: %v\n", filePath, err)
		os.Exit(1)
	}
	defer file.Close()

	if err := t.Execute(file, problem); err != nil {
		fmt.Printf("Failed to write to file %s: %v\n", filePath, err)
		os.Exit(1)
	}
}
